ingestion-sink/src/test/java/com/mozilla/telemetry/ingestion/SinkIntegrationTest.java

```diff
@@ -0,0 +1,146 @@
+    project = bqHelper.getOptions().getProjectId();
+    dataset = RemoteBigQueryHelper.generateDatasetName();
+    bigquery.create(DatasetInfo.newBuilder(dataset).build());
+  }
+
+  /** Remove all resources we created in BigQuery. */
+  @After
+  public void deleteBigQueryDataset() {
+    RemoteBigQueryHelper.forceDelete(bigquery, dataset);
+  }
+
+  @Rule
+  public final EnvironmentVariables environmentVariables = new EnvironmentVariables();
+
+  @Rule
+  public final Timeout timeout = new Timeout(30, TimeUnit.SECONDS);
+
+  private static final TimePartitioning TIME_PARTITIONING = TimePartitioning
+      .newBuilder(TimePartitioning.Type.DAY).setField("submission_timestamp").build();
+
+  private static final Clustering CLUSTERING = Clustering.newBuilder()
+      .setFields(ImmutableList.of("submission_timestamp")).build();
+
+  @Test
+  public void canSinkRawMessages() throws Exception {
+    StandardTableDefinition tableDef = StandardTableDefinition
+        .of(Schema.of(Field.of("document_id", LegacySQLTypeName.STRING),
+            Field.of("document_type", LegacySQLTypeName.STRING),
+            Field.of("document_version", LegacySQLTypeName.STRING),
+            Field.of("submission_timestamp", LegacySQLTypeName.TIMESTAMP)))
+        .toBuilder().setTimePartitioning(TIME_PARTITIONING).setClustering(CLUSTERING).build();
+
+    String submissionTimestamp = ZonedDateTime.now(ZoneOffset.UTC)
+        .format(DateTimeFormatter.ISO_DATE_TIME);
+    Map<String, String> documentIds = ImmutableMap.of("1", UUID.randomUUID().toString());
+    List<String> versions = ImmutableList.of("1", "2");
+    versions.forEach(version -> {
+      String table = "test_v" + version;
+      TableId tableId = TableId.of(dataset, table);
+      bigquery.create(TableInfo.newBuilder(tableId, tableDef).build());
+      PubsubMessage.Builder message = PubsubMessage.newBuilder()
+          .setData(ByteString.copyFrom("test".getBytes())).putAttributes("document_type", "test")
+          .putAttributes("document_version", version)
+          .putAttributes("submission_timestamp", submissionTimestamp);
+      if (documentIds.containsKey(version)) {
+        message.putAttributes("document_id", documentIds.get(version));
+      }
+      try {
+        publisher.publish(message.build()).get();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    });
+
+    environmentVariables.set("INPUT_SUBSCRIPTION", subscriptionName.toString());
+    environmentVariables.set("OUTPUT_TABLE",
+        project + "." + dataset + ".${document_type}_v${document_version}");
+
+    CompletableFuture<Void> main = CompletableFuture.runAsync(() -> Sink.main(new String[] {}));
+
+    String query = "SELECT * REPLACE (FORMAT_TIMESTAMP('%FT%R:%E*SZ', submission_timestamp) AS "
+        + "submission_timestamp), _TABLE_SUFFIX AS table FROM `" + project + "." + dataset + ".*`";
+    AtomicReference<List<List<String>>> actual = new AtomicReference<>();
```

Comment:
I don't think there's any need for an AtomicReference here. We're mutating `actual` within the lambda expression, but not assigning a new value, so there shouldn't be any compiler errors to worry about.

Reply:
it doesn't compile if i unnest it because two lines down I make a new list (`actual.set(new LinkedList<>());`), and variables referenced in a lambda "have to be final or effectively final".

Reply:
Missed that 👍